google/cloud/storage/internal/grpc/object_request_parser.cc

```diff
@@ -377,30 +377,29 @@ StatusOr<google::storage::v2::ReadObjectRequest> ToProto(
   if (request.HasOption<storage::Generation>()) {
     r.set_generation(request.GetOption<storage::Generation>().value());
   }
-  if (request.HasOption<storage::ReadRange>()) {
+  auto const offset = request.GetOption<storage::ReadFromOffset>().value_or(0);
+  if (!request.HasOption<storage::ReadRange>()) {
+    r.set_read_offset(offset);
+  } else {
     auto const range = request.GetOption<storage::ReadRange>().value();
-    r.set_read_offset(range.begin);
-    r.set_read_limit(range.end - range.begin);
+    auto const merged_offset = (std::max)(range.begin, offset);
+    auto const merged_limit = range.end - merged_offset;
+    if (merged_limit <= 0) {
+      return internal::InvalidArgumentError(
+          absl::StrCat("Invalid gRPC range offset=", merged_offset,
+                       ", limit=", merged_limit, ") from offset=", offset,
+                       " range=[", range.begin, ",", range.end, ")"),
```

Comment:
Check your brackets. I see two closing parentheses and one opening square bracket.